main/src/main/scala/sbt/Defaults.scala

```diff
@@ -363,7 +363,14 @@ object Defaults extends BuildCommon {
   lazy val outputConfigPaths = Seq(
     classDirectory := crossTarget.value / (prefix(configuration.value.name) + "classes"),
     target in doc := crossTarget.value / (prefix(configuration.value.name) + "api")
+  ) ++ inTask(compile)(
```

Comment:
No if you look at #3542 it talks about creating a `compile:clean`, not `compile::clean`, i.e a `clean` scoped to the `Compile` configuration, not to the `compile` task.

Reply:
I changed that in the ticket and advised Alexey to do it this way. I think that scoping in the task makes more sense because you may be interested in deleting test class files as well.

Reply:
> I think that scoping in the task makes more sense because you may be interested in deleting test class files as well.

I agree, and `test:clean` should do that.

Reply:
I'm fine with either way. I think scoping by task makes sense if there will be more variants of `clean`. But if it addresses only compilation task, I agree with @dwijnand that it's better to scope it by configuration. Because `compile::clean` gives a false impression that it should magically work with other tasks (cleaning their specific artifacts).

Reply:
> Because `compile::clean` gives a false impression that it should magically work with other tasks (cleaning their specific artifacts).

Exactly.

Reply:
Most things about sbt's scopes are all about compilation/classpath.

Reply:
Hmm, `test:clean` is not the same to `test:compile::clean`. The latter is explicit and portrays better the original intent: only deleting class files. `test:clean` already works now and is not clear about what it should clean, so I strongly favor `test:compile::clean`.
  
>  Because compile::clean gives a false impression that it should magically work with other tasks (cleaning their specific artifacts).
  
Well, this already happens with `test:clean`. You would expect sbt to only delete things generated by the `Test` configuration, but it will delete things generated by `Compile` and `Test`.

Reply:
> `test:clean` already works now and is not clear about what it should clean

> Well, this already happens with `test:clean`. You would expect sbt to only delete things generated by the `Test` configuration, but it will delete things generated by `Compile` and `Test`.

I'm not sure what you mean by "works" then. Currently `test:clean` just delegates to `clean`. Thus #3542.

There is actually a sys property to disable that delegation and I was playing with it the other day: https://github.com/sbt/sbt/blob/05c2c506b2b218d8ed8befcef843951391c07be6/main/src/main/scala/sbt/internal/Act.scala#L446

It's interesting because it mean that `test:clean` wouldn't run because it wouldn't delegate to `clean`, but `test` would continue to delegate to `test:test` and therefore run. I want to revisit it for sbt 2.

Reply:
> I'm not sure what you mean by "works" then. Currently test:clean just delegates to clean. Thus #3542.
  
Yup, I agree. This is why I said:
  
>  this already happens with test:clean. You would expect sbt to only delete things generated by the Test configuration, but it will delete things generated by Compile and Test.
  
That sys property is interesting, didn't know it.
  
I'm trying to think how we solve this puzzle to get the intended behaviour in `test:compile::clean`.

Reply:
It turns out we've been missing that the actual implementation does only delete test class files, so we're good. Turning the `test` into the following makes the test pass:
```
> test:compile
$ exists target/scala-2.12/resolution-cache/
$ exists target/streams/
$ exists target/scala-2.12/classes/A.class
$ exists target/scala-2.12/test-classes/B.class

> test:compile::clean
$ exists target/scala-2.12/resolution-cache/
$ exists target/streams/
# it should clean only test class files
$ exists target/scala-2.12/classes/A.class
$ absent target/scala-2.12/test-classes/B.class

> compile:compile::clean
$ exists target/scala-2.12/resolution-cache/
$ exists target/streams/
# it should clean only compile class files
$ absent target/scala-2.12/classes/A.class
$ absent target/scala-2.12/test-classes/B.class
```

Reply:
> There is actually a sys property to disable that delegation and I was playing with it the other day

Wow! I just told @jvican yesterday that delegation is confusing and I'd like sbt to fail when I mix up scopes and get a delegated value without realising it.



> It turns out we've been missing that the actual implementation does only delete test class files, so we're good.

👍 done in f2aab7849472611cad1285793c0b455e246f0024